### DISCUSSION TITLE
파일별 코드 내용 통일

### DISCUSSION THREAD

**Discussion by devnproyj22** [weaverse-techtide/weaverse-backend#98] [General]
# 파일별 코드 내용 통일

* 파일별로 의견 올려주세요.

### serializers.py
- ModelSerializers 상속
- 직렬화: to_representation 메서드 오버라이딩
- 역직렬화: create, update
### views.py
- GenericAPIView + Mixin 상속

**Comment by devnproyj22** [comment]
## 장고 RF 의 뷰 구현 방법
DRF는 API 뷰를 구현하는 여러 방법을 제공하며, 각 방법은 추상화 수준과 코드 복잡성 측면에서 차이가 있습니다.

[V] CBV:  하나의 클래스가 하나의 URL 에 대한 모든 요청을 처리한다. 
[  ] FBV: 하나의 함수가 각 요청을 처리하고 하나의 URL에 대해 여러 함수가 처리한다.
![image](https://github.com/user-attachments/assets/2062231b-f051-43f5-a7d4-5ce6e5fb0d51)

### 소개

### 기준
* 코드 복잡성(중복)
* 추상화 수준
* 유연성

### 👍 선택

단순한 CRUD 작업-> Generic APIView나 ViewSet이 적합
복잡한 로직이나 세밀한 제어 필요-> APIView나 APIView+Mixin이 적합
중간 수준의 커스터마이징-> GenericAPIView + Mixin가 좋은 균형점

### APIView 상속
가장 기본적인 방법: 장고의 View와 유사 (serializer를 추가해 처리해준것뿐)
-> 각 요청 메서드마다 직접 serializer 처리를 합니다.
각 HTTP 메서드(GET, POST 등)를 개별적으로 정의합니다.
-> 각 메서드 간 코드 중복이 발생합니다.
복잡한 로직이 필요한 경우 적합합니다. 

#### 코드 
| 구성 요소 | 들어갈 내용 |
| :----------: | :------------ |
| Serializer | 일반적으로 하나의 serializer 클래스가 필요합니다. |
| URLs | 각 뷰 클래스마다 하나의 독립적인 URL 패턴이 필요합니다. |
| Permission | permission_classes 속성, 뷰의 구현 방법과 관계없이 권한 요구사항에 따라 하나 이상의 권한 클래스를 사용합니다. |
| View |

### APIView + Mixin 동시 상속
APIView의 유연성과 Mixin의 재사용성을 결합하여 작성합니다.
-> 공통 CRUD 작업을 Mixin으로 처리하면서, 필요한 경우 세부적인 커스터마이징이 가능합니다.
-> APIView보다 코드 중복이 줄어들지만, 여전히 HTTP 메서드와 Mixin 메서드를 연결해야 합니다.
-> APIView에서 발생하는 여러 serializer에 대한 반복 중복 문제를 보완하는 방법입니다. 
->  queryset 과 serializer_class 를 지정해주기만 하면 나머지는 상속받은 Mixin 과 연결해주기만 하면 됩니다.

#### 코드 
| 구성 요소 | 들어갈 내용 |
| :----------: | :------------ |
| Serializer | 일반적으로 하나의 serializer 클래스가 필요합니다. |
| URLs | 각 뷰 클래스마다 하나의 독립적인 URL 패턴이 필요합니다. |
| Permission | APIView와 동일 |
| View |

###  GenericAPIView + Mixin 동시 상속
GenericAPIView는 공통적으로 사용되는 속성(queryset, serializer_class 등)을 제공합니다.
Mixin과 결합하여 CRUD 작업을 더 간단하게 구현할 수 있습니다.
APIView + Mixin보다 추상화 수준이 높아 코드가 간결해집니다.

#### 코드 
| 구성 요소 | 들어갈 내용 |
| :----------: | :------------ |
| Serializer | serializer_class 속성으로 하나의 serializer 클래스를 지정합니다. |
| URLs | 각 뷰 클래스마다 하나의 독립적인 URL 패턴이 필요합니다. |
| Permission | APIView와 동일 |
| View |

### Generic APIView 상속
Mixin + GenericAPIView의 조합을 미리 구현해 놓은 클래스들입니다.
가장 일반적인 CRUD 작업에 대해 미리 정의된 뷰를 제공합니다.
Mixin에서 발생하는 여러 Mixin 상속으로 가독성 떨어지는 문제를 보완하는 방법입니다.
-> Mixin을 상속한 새로운 Generics 클래스를 사용합니다. 
-> Mixin과 결합하여 CRUD 작업을 더 간단하게 구현할 수 있습니다.
코드가 매우 간결해지지만, 커스터마이징의 유연성은 상대적으로 낮아집니다.

```
generics.CreateAPIView : 생성
generics.ListAPIView : 목록
generics.RetrieveAPIView : 조회
generics.DestroyAPIView : 삭제
generics.UpdateAPIView : 수정
generics.RetrieveUpdateAPIView : 조회/수정
generics.RetrieveDestroyAPIView : 조회/삭제
generics.ListCreateAPIView : 목록/생성
generics.RetrieveUpdateDestroyAPIView : 조회/수정/삭제
```

#### 코드 
| 구성 요소 | 들어갈 내용 |
| :----------: | :------------ |
| Serializer | serializer_class 속성으로 하나의 serializer 클래스를 지정합니다. |
| URLs | 각 뷰 클래스마다 하나의 독립적인 URL 패턴이 필요합니다. |
| Permission | APIView와 동일 |
| View |

### ViewSet 상속
관련된 뷰들을 하나의 클래스로 그룹화합니다.
Router를 사용하여 URL 설정을 자동화할 수 있습니다.
CRUD 작업에 대한 표준 구현을 제공하며, 필요한 경우 커스터마이징도 가능합니다.
Generics APIView에서 발생하는 공통 queryset 과 serializer_class를 따로가 아닌 한 번에 처리해주는 방법입니다. 
코드 중복을 크게 줄이고 일관된 API 구조를 만들 수 있지만, 학습 곡선이 있을 수 있습니다.
ViewSet은 CBV 가 아닌 헬퍼클래스로 두 가지 종류가 있습니다.
일반적인 CBV 와 다르게 as_view를 통해서 뷰를 연결하지 않고, router를 사용해서 연결하였습니다.
ViewSet은 하나의 뷰가 아닌 set, 여러 개의 뷰를 만들 수 있는 확장된 CBV입니다.

#### 코드 
| 구성 요소 | 들어갈 내용 |
| :----------: | :------------ |
| Serializer | serializer_class 속성으로 하나의 serializer 클래스를 지정합니다. 필요한 경우 action별로 다른 serializer를 사용할 수 있습니다.
| URLs | Router를 사용하여 여러 URL 패턴을 자동으로 생성합니다. |
| Permission | permission_classes 속성, @action, get_permissions 메서드 오버라이딩 |
| View |

**Comment by Yoo117** [comment]
## 1. test/
테스트 파일은 저번에 합의했던 대로 test 폴더를 만들어서 개인적으로 구현하는 식으로 하면 좋을 것 같습니다. 

## 2. Serializer
serializer 같은 경우는 ModelSerializer에 직렬화, 역직렬화 기능이 모두 포함되어 있다고 합니다. 또 ModelSerializer의 메서드로 직렬화, 역직렬화 내용을 커스텀 할 수 있다고 합니다. 굳이 번거롭게 직접 Serializer의 직렬화/역직렬화를 구현할 필요는 없다고 생각합니다.

## 3. View
사용 방식이 프로젝트의 규모, 상황별, 로직의 복합성, 팀원의 선호도에 따라 달라진다고 하더라구요.

**예시) 상황별에 따라 View 사용 구분**
- 표준 CRUD: Generic Views나  ViewSets
- 복잡한 비즈니스 로직: APIView
- 커스텀 동작 + 재사용성:  Mixins + Generic Views

일반적인 CRUD 기능 구현엔 Generic Views나 ViewSets, 조금 커스텀이 필요한 view를 구현할 필요가 있다면 Mixins나 APIView를 사용하는 편이 좋을 것 같습니다. 그러나 APIView는 하나하나 다 구현해야 하는 불편함이 따르기에, 결국 홍광님이 현재 사용하고 계신 Generic Views(일반적인 CRUD 구현) + Mixins(커스텀 view 구현) 방식이 괜찮을 것 같습니다.

**Comment by AlbertImKr** [comment]
Mixins + Generic Views 사용하기로 결정했습니다.
